codec-http2/src/main/java/io/netty/handler/codec/http2/ChildChannel.java

```diff
@@ -0,0 +1,252 @@
+        }
+    };
+
+    private boolean closed;
+    private boolean readInProgress;
+
+    public ChildChannel(Channel parent) {
+        super(parent);
+    }
+
+    @Override
+    public ChannelMetadata metadata() {
+        return METADATA;
+    }
+
+    @Override
+    public ChannelConfig config() {
+        return config;
+    }
+
+    @Override
+    public boolean isOpen() {
+        return !closed;
+    }
+
+    @Override
+    public boolean isActive() {
+        return !closed;
+    }
+
+    @Override
+    protected AbstractUnsafe newUnsafe() {
+        return new ChildUnsafe();
+    }
+
+    @Override
+    protected boolean isCompatible(EventLoop loop) {
+        return loop instanceof SingleThreadEventLoop;
+    }
+
+    @Override
+    protected SocketAddress localAddress0() {
+        return parent().localAddress();
+    }
+
+    @Override
+    protected SocketAddress remoteAddress0() {
+        return parent().remoteAddress();
+    }
+
+    @Override
+    protected void doRegister() throws Exception {
+        ((SingleThreadEventLoop) eventLoop().unwrap()).addShutdownHook(shutdownHook);
+    }
+
+    @Override
+    protected void doBind(SocketAddress localAddress) throws Exception { }
+
+    @Override
+    protected void doDisconnect() throws Exception {
+        doClose();
+    }
+
+    @Override
+    protected void doClose() throws Exception {
+        closed = true;
+    }
+
+    @Override
+    protected void doDeregister() throws Exception {
+        // Just remove the shutdownHook as this Channel may be closed later or registered to another
+        // EventLoop
+        ((SingleThreadEventLoop) eventLoop().unwrap()).removeShutdownHook(shutdownHook);
+    }
+
+    @Override
+    protected void doBeginRead() throws Exception {
+        if (readInProgress) {
+            return;
+        }
+
+        if (inboundBuffer.isEmpty()) {
+            readInProgress = true;
+            return;
+        }
+
+        final InternalThreadLocalMap threadLocals = InternalThreadLocalMap.get();
+        final Integer stackDepth = threadLocals.localChannelReaderStackDepth();
+        if (stackDepth < MAX_READER_STACK_DEPTH) {
+            threadLocals.setLocalChannelReaderStackDepth(stackDepth + 1);
+            try {
+                readTask.run();
+            } finally {
+                threadLocals.setLocalChannelReaderStackDepth(stackDepth);
+            }
+        } else {
+            eventLoop().execute(readTask);
+        }
+    }
+
+    @Override
+    protected final void doWrite(ChannelOutboundBuffer in) throws Exception {
+        if (closed) {
+            throw new ClosedChannelException();
+        }
+
+        EventLoop parentLoop = parent().eventLoop();
+
+        if (parentLoop.inEventLoop()) {
+            try {
+                for (;;) {
+                    Object msg = in.current();
+                    if (msg == null) {
+                        break;
+                    }
+                    parentWrite(msg);
```

Comment:
A return status only solves one of the two cases. If we are using a different event loop we have to complete the promise before we know whether the parent supports it or has any error. We could special-case that one check, because the supported messages is fixed, but I wonder what other sorts of problems we will have.

Most solutions seem to involve avoiding or modifying ChannelOutboundBuffer.